Configuration option to disable transition periods

bx80 opened the pull request:
### Description:

Fixes #18332 

 Adds an option to specify the max allowed period for transitions, either globally or per site

```
[Transitions]
max_period_allowed = "month"
```

```
[Transitions_1]
max_period_allowed = "week"
```
Site level settings override the general settings.
All period smaller than the max period are allowed, so month will also allow week and day.
Ranges with more days than the max period will not be allowed, eg. an 8 day range isn't allowed if the `max_period_allowed = "week"`, but a 7 day range would be ok.

Row transitions obey the same rules, but I haven't hidden the row transition action icon if the period is not allowed as it might be confusing to have to icon disappear entirely without any explanation. It's also a bit awkward having the datatable row action registration methods make an async API call each time to check period availability - open to suggestions on how to do this efficiently if we really want to go ahead and hide the row action :slightly_smiling_face: 

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

EreMaijala committed:
Add option to specify the max number of websites to process by archiver.

This makes it possible to limit the process lifetime to avoid excessive memory usage (possibly due to PHP bugs such as 79519).

EreMaijala committed:
Add an option to limit the number of archives to process.

EreMaijala committed:
Clean up a whitespace change.

EreMaijala committed:
Make --max-archives-to-process a hard limit.

Commit:
Added config option to disable transition periods

sgiehl commented on plugins/Transitions/API.php:
```diff
@@ -656,4 +660,63 @@ protected function processTransitionsToFollowingPages($type, $actions)
+     */
+    private function getPeriodAllowed($period, $idSite, $date) : bool
+    {
+        $generalMaxPeriodAllowed = Config::getInstance()->Transitions['max_period_allowed'];
```
This will throw a notice, as the `Transitions` section isn't available by default.

bx80 replied:
I've added some extra null checks.

sgiehl commented on plugins/Transitions/API.php:
```diff
@@ -656,4 +660,63 @@ protected function processTransitionsToFollowingPages($type, $actions)
+        $generalMaxPeriodAllowed = Config::getInstance()->Transitions['max_period_allowed'];
+        $sectionName = 'Transitions_' . $idSite;
+        $section = Config::getInstance()->$sectionName;
+        $siteMaxPeriodAllowed = $section['max_period_allowed'];
```
Same here

sgiehl commented on plugins/Transitions/javascripts/transitions.js:
```diff
@@ -1586,8 +1586,8 @@ Piwik_Transitions_Ajax.prototype.callApi = function (method, params, callback) {
 
                 if (typeof Piwik_Transitions_Translations == 'undefined') {
                     self.callApi('Transitions.getTranslations', {}, function (response) {
-                        if (typeof response[0] == 'object') {
-                            Piwik_Transitions_Translations = response[0];
+                        if (typeof response == 'object') {
+                            Piwik_Transitions_Translations = response;
```
Why was this change required? Wasn't that working correctly before?

bx80 replied:
This wasn't working and the translations were not being loaded. The API returns an object with the translation strings but the response handler is checking for a sub-object for some reason.

justinvelluppillai replied:
Probably related to a change a while back where we removed the wrapping array from json responses?

See https://github.com/matomo-org/matomo/commit/66bcb98556e5b92b7b6187a5a904e13a408d0ed6#diff-287d49d80349087bbbaa1186c9f825689026f878c9bad60f46a735b65011acb2R64

and https://github.com/matomo-org/matomo/pull/18136

sgiehl reviewed (changes requested):
Left some comments. Besides that, according to the original issue it would be good to hide the transitions row action when it's disabled for the current period.
Guess to achieve that, the simples solution would be to pass the config setting as a global javascript variable using the even `Template.jsGlobalVariables`, and then you can check that variable in the row action method that returns it's availability: https://github.com/matomo-org/matomo/blob/c83fe5a1beeb0dd2e2c17448e6cf15c00183f40f/plugins/Transitions/javascripts/transitions.js#L125-L137

Commit:
Config section checks, added global config setting

Commit:
Revert global.ini.php setting

Commit:
Hide transitions row action if period is not allowed

Commit:
Added system test

bx80 commented:
@sgiehl Thanks, `Template.jsGlobalVariables` was the pointer I was needed. I've added code to disable the transitions row action if the period isn't allowed. Also added a system test to make sure the config setting results in the correct exception.

Commit:
Add default values for requests vars

Commit:
Merge commit 'refs/pull/18326/head' of github.com:matomo-org/matomo into 4.x-dev

sgiehl commented on plugins/Transitions/Transitions.php:
```diff
@@ -55,4 +59,24 @@ public function getClientSideTranslationKeys(&$translationKeys)
+        }
+
+        $out .= "      
+        piwik.transitionsPeriodAllowed = $allowed;\n
```
This won't work properly. I guess you will actually need to pass over the configured allowed period to javascript and do the check in javascript as well.
When switching the period/date in the date selector the page won't be fully reloaded, and so the global variables won't be refreshed. So it currently always uses the initial state matomo was opened with.

bx80 replied:
Good point :+1:  I've changed it to pass over the max period allowed config setting and replicate the API check in Javascript. The transitions row action is now disabled/enabled when report dates are changed (without reloading)

Commit:
Do allowed period checks for transition row actions in javascript

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into 4.x-dev

Commit:
Merge branch '4.x-dev' into m-18332-disable-transition-periods

bx80 committed:
built vue files

sgiehl commented on plugins/CoreHome/vue/src/Periods/Range.ts:
```diff
@@ -176,6 +176,10 @@ export default class RangePeriod {
   }
+
+  getDayCount(): number {
+    return Math.ceil((this.endDate.getTime() - this.startDate.getTime()) / (1000 * 3600 * 24));
```
I think that is actually not accurate. It might always return one day less than expected.
Let's assume a two days range from 2021-11-24 to 2021-11-25. As we use `00:00:00` as time the result of `getDayCount` will be `1`, as there is actually a difference of one day. But the reports will show two days. So guess we always need to add `1`. The same also applies for the php code.

bx80 replied:
Fixed. I was incorrectly assuming the range had start and end times, but obviously it doesn't.

sgiehl commented on core/Period/Range.php:
```diff
@@ -498,6 +498,16 @@ public static function getDateXPeriodsAgo($subXPeriods, $date = false, $period =
+     */
+    public function getDayCount()
+    {
+         return self::getNumDaysDifference($this->getDateStart(), $this->getDateEnd());
```
See the other comment

sgiehl commented on plugins/Transitions/API.php:
```diff
@@ -656,4 +660,80 @@ protected function processTransitionsToFollowingPages($type, $actions)
+     *
+     * @return string|null
+     */
+    public function getPeriodAllowedConfig($idSite) : ?string
```
In which case can this method actually return null? If no config setting is set it will return `all`, in all other cases set config value that is set, which is always not `null`

sgiehl replied:
Another note: When adding public methods to API classes, please always keep in mind, that all those methods will be listed in the API docs. We should avoid adding to many methods that are actually useless for a user or are for internal use only. Personally I wouldn't consider this method as much useful, as everything that a user might need is provided by the second method below.
I guess you made this method public to be able to use it in the Transitions plugin class. In that case maybe simply move this method to the Transitions plugin class as a static one, and call it from within the other API method.

bx80 replied:
Good point, I'll make a note to avoid adding any public API methods unless they'd actually be useful for someone using the API.

sgiehl commented on plugins/Transitions/API.php:
```diff
@@ -656,4 +660,80 @@ protected function processTransitionsToFollowingPages($type, $actions)
+        }
+
+        // Site setting overrides general, if it exists
+        return ($siteMaxPeriodAllowed ? $siteMaxPeriodAllowed : $generalMaxPeriodAllowed);
```
could be shortened to

```suggestion
        return $siteMaxPeriodAllowed ?? $generalMaxPeriodAllowed;
```

sgiehl commented on plugins/Transitions/API.php:
```diff
@@ -656,4 +660,80 @@ protected function processTransitionsToFollowingPages($type, $actions)
+     *
+     * @return bool
+     */
+    public function getPeriodAllowed($period, $idSite, $date) : bool
```
We have quite a lot methods using the order `$idSite, $period, $date`, might be good to change it here as well to have some kind of consistency.

sgiehl replied:
Also I would suggest to rename this method to `isPeriodAllowed`, as the method checks if the given period is allowed and doesn't return the allowed period

bx80 replied:
Makes sense, done.

sgiehl commented on plugins/Transitions/tests/System/TransitionsMaxAllowedPeriodTest.php:
```diff
@@ -0,0 +1,45 @@
+ * @group TransitionsMaxAllowedPeriodTest
+ * @group Plugins
+ */
+class TransitionsMaxAllowedPeriodTest extends IntegrationTestCase
```
Integration tests should be located at `tests/Integration/`

sgiehl commented on plugins/Transitions/tests/System/TransitionsMaxAllowedPeriodTest.php:
```diff
@@ -0,0 +1,45 @@
+    }
+}
+
+TransitionsMaxAllowedPeriodTest::$fixture = new SomeVisitsManyPageviewsWithTransitions();
```
Integration tests don't really need fixture
```suggestion
```

sgiehl commented on plugins/Transitions/tests/System/TransitionsMaxAllowedPeriodTest.php:
```diff
@@ -0,0 +1,45 @@
+        $this->api = API::getInstance();
+    }
+
+    public function test_ShouldThrowException_IfPeriodNotAllowed()
+    {
+        // Attempt to get transition data for a week when the max period is set to a day
+        Config::setSetting('Transitions', 'max_period_allowed', 'day');
+        $this->expectException(\Exception::class);
+        $this->expectExceptionMessage('PeriodNotAllowed');
+        $this->api->getTransitionsForAction('http://example.org/page/one.html', 'url', self::$fixture->idSite, 'week', self::$fixture->dateTime);
+    }
```
Would be good to add some more tests to test other periods and range edge cases. Also maybe some tests that valid requests don't return an exception

bx80 replied:
I've added some tests for valid and invalid combinations, that ranges that are larger than periods throw exceptions and that per set config values override the general config setting.

sgiehl reviewed (changes requested):
Left a couple of comments.

Commit:
Code tidy up, fix for range get day count bug, improved tests

Commit:
Build vue

bx80 committed:
built vue files

sgiehl reviewed (approved):
Looked through the code and tested everything locally. Works as expected 👍 
You could add an additional UI test with the period that should be shown disabled, but that's a nice to have. Besides that, feel free to merge.

Commit:
Added UI test for disabled period

Commit:
Merge remote-tracking branch 'origin/m-18332-disable-transition-periods' into m-18332-disable-transition-periods
